deps/db/src/logseq/db/rules.cljc

```diff
@@ -2,62 +2,63 @@
-     [?c :block/parent ?p]]
-    [(child ?p ?c)
-     [?t :block/parent ?p]
-     (child ?t ?c)]
-
-  ;; rule "namespace" is optimized for child node -> node of upper namespace level nesting queries
-    [(namespace ?p ?c)
-     [?c :block/namespace ?p]]
-    [(namespace ?p ?c)
-     [?t :block/namespace ?p]
-     (namespace ?t ?c)]
-
-    ;; Select rules carefully, as it is critical for performance.
-    ;; The rules have different clause order and resolving directions.
-    ;; Clause order Reference:
-    ;; https://docs.datomic.com/on-prem/query/query-executing.html#clause-order
-    ;; Recursive optimization Reference:
-    ;; https://stackoverflow.com/questions/42457136/recursive-datalog-queries-for-datomic-really-slow
-    ;; Should optimize for query the decendents of a block
-    ;; Quote:
-    ;; My theory is that your rules are not written in a way that Datalog can optimize for this read pattern - probably resulting in a traversal of all the entities. I suggest to rewrite them as follows:
-    ;; [[(ubersymbol ?c ?p)
-    ;;   (?c :ml/parent ?p)]
-    ;;  [(ubersymbol ?c ?p)
-    ;;   ;; we bind a child of the ancestor, instead of a parent of the descendant
-    ;;   (?c1 :ml/parent ?p)
-    ;;   (ubersymbol ?c ?c1)]]
-
-    ;; This way of writing the ruleset is optimized to find the descendants of some node. The way you originally wrote it is optimized to find the anscestors of some node.
-
-    ;; from https://stackoverflow.com/questions/43784258/find-entities-whose-ref-to-many-attribute-contains-all-elements-of-input
-    ;; Quote:
-    ;; You're tackling the general problem of 'dynamic conjunction' in Datomic's Datalog.
-    ;; Write a dynamic Datalog query which uses 2 negations and 1 disjunction or a recursive rule
-    ;; Datalog has no direct way of expressing dynamic conjunction (logical AND / 'for all ...' / set intersection).
-    ;; However, you can achieve it in pure Datalog by combining one disjunction
-    ;; (logical OR / 'exists ...' / set union) and two negations, i.e
-    ;; (For all ?g in ?Gs p(?e,?g)) <=> NOT(Exists ?g in ?Gs, such that NOT(p(?e, ?g)))
-
-    ;; [(matches-all ?e ?a ?vs)
-    ;;  [(first ?vs) ?v0]
-    ;;  [?e ?a ?v0]
-    ;;  (not-join [?e ?vs]
-    ;;            [(identity ?vs) [?v ...]]
-    ;;            (not-join [?e ?v]
-    ;;                      [?e ?a ?v]))]
-    ])
+  {:namespace
+   '[[(namespace ?p ?c)
+      [?c :block/namespace ?p]]
+     [(namespace ?p ?c)
+      [?t :block/namespace ?p]
+      (namespace ?t ?c)]]
+
+   :alias
+   '[[(alias ?e2 ?e1)
+      [?e2 :block/alias ?e1]]
+     [(alias ?e2 ?e1)
+      [?e1 :block/alias ?e2]]
+     [(alias ?e1 ?e3)
+      [?e1 :block/alias ?e2]
+      [?e2 :block/alias ?e3]]
+     [(alias ?e3 ?e1)
+      [?e1 :block/alias ?e2]
+      [?e2 :block/alias ?e3]]]})
+
+;; Rules writing advice
+;; ====================
+;; Select rules carefully, as it is critical for performance.
+;; The rules have different clause order and resolving directions.
```

Comment:
What clause order and resolving direction does a rule have?
The doc below describes clause order of a `where`.

Reply:
Rules and `where` clauses are exactly the same. The rules will be applied in the `where` clauses
Reference:
https://docs.datomic.com/on-prem/query/query.html#rules
https://github.com/logseq/logseq/blob/a4ee4c48f88f95f1840a622daac72e7b6e45a58b/src/main/frontend/db/model.cljs#L1588

Reply:
Then I don't get what that mean :thinking: 
> The rules have *different* clause order and resolving directions.

Reply:
Humm. In short, if you use a wrong order of the clauses, the query engine would have to iterate all the "possible values" of the variables, and it will be very slow.
I think the common practice here should be to "bind" / "lock-in" the actual value of variables as early as possible to minimize the "possible values" spaces.
Ref:
https://docs.datomic.com/on-prem/query/query-executing.html#clause-order

Reply:
Feel free to add your comments in code so the followers can share our knowledge :)

Reply:
Thanks for the explanation. That bit is clear.
What I don't get is how clause order and resolving direction are *different* for rules.
So far it seems there is nothing special in behaviour of rules.

Reply:
@andrewzhurov Hopefully the original performance issue of this optimization to fix could help your understanding:
Issue: https://github.com/logseq/logseq/issues/4362
PR: https://github.com/logseq/logseq/pull/4424

Reply:
That PR seem to add performance boost by resolving from child to parent.
I guess that is what's been meant by 'different resolving direction' then.
I guess 'different clause order' means something along those lines too.

My confusion with this comment is that it's written in a section which I considered to be "general rule-writing tips", whereas this tip seem to refer specificly to some of the written rules.
My understanding from a glance of "rules have different resolution direction" is that if we write
```
clause 1
clause 2
clause 3
```
they will be executed in order: clause 3, clause 2, clause 1.
Akin to how 'strangeley' `comp` behaves with transducers.
Whereas in fact this tip seems to refer to some specific use-cases we have and requires some context to get it.

I, as for somebody with little project context, would benefit from more explicit wording along lines:
"It's performant to query data with most constricting patterns first, keep an eye for it, sometimes it's efficient to get a "child" and ask for it's "parent""

Reply:
Humm, the order should be `clause 1`, `clause 2`, `clause 3`. The comment here means rule `parent` and rule `child` are having different orders.

Reply:
Yup, I got that after you tuned me up on the context, thanks.:)
Perhaps explaining it in the tip line would make it less confusion-prone for bystanders such as I.